### ISSUE TITLE
[enhancement] Validate Autosubmit configuration

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#871]
Hi,

Yesterday after a meeting @mcastril and I were chatting about Autosubmit and other projects at BSC, and at some point we discussed validating YAML files.

I mentioned that in [cwltool](https://github.com/common-workflow-language/cwltool), CWL workflows are written in YAML, and the validation (also code generation, HTML documentation, and more) is handled by a library they created, called [Schema Salad](https://github.com/common-workflow-language/schema_salad/).

This library was created after they tried to use JSON-LD but found features missing from the specification (I believe they discussed it directly with the specification working group).

Schema Salad is able to, given a schema written in JSON or YAML, validate that an input YAML is valid or not, based on the schema structure, types, and attributes.

Maybe something that could be useful for Autosubmit 4 too?

-Bruno

**Comment by kinow** [comment]
For example, given my current AS4 platforms file:

```yaml
Platforms:
  local:
    TYPE: ps
    HOST: localhost
    PROJECT: test
    USER: kinow
    QUEUE: debug
    SCRATCH_DIR: '/tmp/'
    ADD_PROJECT_TO_HOST: False
    MAX_WALLCLOCK: '48:00'
    TEMP_DIR: '/tmp/'
  MaReNoStRuM4:
    TYPE: slurm
    HOST: mn1.bsc.es
    PROJECT: bsc32
    USER: bsc32070
    QUEUE: debug
    SCRATCH_DIR: /gpfs/scratch
    ADD_PROJECT_TO_HOST: False
    MAX_WALLCLOCK: '48:00'
    TEMP_DIR: ''
  marenostrum_archive:
    TYPE: ps
    HOST: dt02.bsc.es
    PROJECT: bsc32
    USER: bsc32070
    SCRATCH_DIR: /gpfs/scratch
    ADD_PROJECT_TO_HOST: False
    TEST_SUITE: False

  transfer_node:
    TYPE: ps
    HOST: dt01.bsc.es
    PROJECT: bsc32
    USER: bsc32070
    ADD_PROJECT_TO_HOST: false
    SCRATCH_DIR: /gpfs/scratch

  transfer_node_bscearth000:
    TYPE: ps
    HOST: bscearth000
    USER: dbeltran
    PROJECT: Earth
    ADD_PROJECT_TO_HOST: false
    QUEUE: serial
    SCRATCH_DIR: /esarchive/scratch

  bscearth000:
    TYPE: ps
    HOST: bscearth000
    USER: dbeltran
    PROJECT: Earth
    ADD_PROJECT_TO_HOST: false
    QUEUE: serial
    SCRATCH_DIR: /esarchive/scratch
  nord3:
    TYPE: SLURM
    HOST: nord1.bsc.es
    PROJECT: bsc32
    USER: bsc32070
    QUEUE: debug
    SCRATCH_DIR: /gpfs/scratch
    MAX_WALLCLOCK: '48:00'

  ecmwf-xc40:
    TYPE: ecaccess
    VERSION: pbs
    HOST: cca
    USER: c3d
    PROJECT: spesiccf
    ADD_PROJECT_TO_HOST: false
    SCRATCH_DIR: /scratch/ms
    QUEUE: np
    SERIAL_QUEUE: ns
    MAX_WALLCLOCK: '48:00'
```

And the following schema:

```yaml
saladVersion: v1.1
$base: "https://bsc.es/autosubmit#"

$namespaces:
  autosubmit: "https://bsc.es/autosubmit#"
  sld: "https://w3id.org/cwl/salad#"

$graph:

- name: Platform
  type: record
  fields:
    id:
      type: string?
      jsonldPredicate: "@id"
    TYPE:
      type:
        type: enum
        symbols:
            - ps
            - pbs
            - ssh
            - slurm
            - SLURM
            - ecaccess
    HOST: string
    PROJECT: string?
    USER: string
    QUEUE: string?
    SERIAL_QUEUE: string?
    SCRATCH_DIR: string
    ADD_PROJECT_TO_HOST: boolean?
    MAX_WALLCLOCK: string?
    TEMP_DIR: string?
    TEST_SUITE: boolean?
    VERSION: string?

- name: PlatformsObject
  doc: |
    The platforms document root...
  type: record
  documentRoot: true
  fields:
    - name: Platforms
      type:
        type: array
        items: Platform
      jsonldPredicate:
        mapSubject: id
```

I can validate the platforms file with `schema-salad-tool`:

```bash
(venv) kinow@ranma:~/Development/python/workspace/schema_salad$ schema-salad-tool schema.yml platforms.yml
/home/kinow/Development/python/workspace/schema_salad/venv/bin/schema-salad-tool Current version: 8.3.20220916115321
Document `platforms.yml` is valid
```

If I change the type of the “local” platform to something like "Ferrari":

```bash
(venv) kinow@ranma:~/Development/python/workspace/schema_salad$ schema-salad-tool schema.yml platforms.yml
/home/kinow/Development/python/workspace/schema_salad/venv/bin/schema-salad-tool Current version: 8.3.20220916115321
While validating document `platforms.yml`:
platforms.yml:1:1: Object `platforms.yml` is not valid because
                     tried `PlatformsObject` but
                       the `Platforms` field is not valid because
platforms.yml:2:3:       item is invalid because
platforms.yml:3:5:         the `TYPE` field is not valid because
                             the value 'ferrari' is not a valid
                             anon.enum_12642360466393a873f791d0f2aad029147df8d3, expected one of 'ps', 'pbs',
                             'ssh', 'slurm', 'SLURM', 'ecaccess'
```

**Comment by kinow** [comment]
I saw today a friend from MetOffice had released a GitHub library about YAML. After reading the docs, I think it does something similar to schema-salad, without some features (extending, interfaces, enums, etc), but with other features (variable substitution).

https://github.com/JCSDA-internal/yamlprocessor

Might be interesting to compare both too. I might drop him a note later to ask if he had heard about schema-salad, and where that library is used.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 7, 2023, 16:38

unassigned @dbeltrankyl

**Comment by kinow** [comment]
Another way of doing validation for JSON & YAML: https://cerfacs.fr/coop/json-schema-for-sci-apps

(p.s.: CERFACS has a lot of interesting articles too on programming, licenses, science, and HPC)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 29, 2023, 09:48

Hello @kinow ,

Maybe this is worth looking into soon because the DestinE users will start to use the workflow and AS4 with little knowledge,

I considered integrating this into autosubmit 4.1 whenever they try to change some parameter in their $expdef/conf folder ( or main.yml for that specific case ) to avoid human issues.

Also, would it be worth including some schema generators for developer users? Through an Autosubmit command,  So for example, in the case of DestinE workflow, define the valid parameters for the PRESET values they have set so the user can't make mistakes

About the integration, it seems a bit tricky...

* If we integrate it file by file, what will happen is that the validator won't work because some parameters will have dummy values like ( %PLATFORMS.%PLATFORM%.USER%: "INTRODUCEYOURUSER" )
* If we integrate it after all files are unified, we must store the unify.yml somewhere ( that is already needed for the GUI ) because the schema needs it, but then we won't know which file has the typo, so we would have to do some post-process to locate it 

However, this seems like a lot of work, and I don't have time, and I think yours is also limited. What do you think?


Edit: Could you have multiple schemas for a single file? To check different parts of it ( per example, developers side and AS variables )

**Comment by kinow** [comment]
>Maybe this is worth looking into soon because the DestinE users will start to use the workflow and AS4 with little knowledge,

Good point, I think you are right.

>I considered integrating this into autosubmit 4.1 whenever they try to change some parameter in their $expdef/conf folder ( or main.yml for that specific case ) to avoid human issues.

Yeah, would be nice to have something like a validate command, that checks your configuration is valid (if used automatically before some commands it could avoid the cases where users may break the configuration and not be able to run commands until they fix it manually, I think).

>Also, would it be worth including some schema generators for developer users? Through an Autosubmit command, So for example, in the case of DestinE workflow, define the valid parameters for the PRESET values they have set so the user can't make mistakes

Schema salad also produces documentation out of the provided schema (it validates, generates docs/HTML, and is also able to produce code that consumes the schema in other languages), but first we would have to confirm if Schema Salad will work with our configuration model (I'm a bit concerned as we have strings that store arrays, like "value,value,value" or "value value value" I think).

>About the integration, it seems a bit tricky...
>
>If we integrate it file by file, what will happen is that the validator won't work because some parameters will have dummy values like ( %PLATFORMS.%PLATFORM%.USER%: "INTRODUCEYOURUSER" )

Good point. Not sure if the schema-salad validation would work before or only after the preprocessing of the configuration, or if we would have to give it the complete YAML file.

>If we integrate it after all files are unified, we must store the unify.yml somewhere ( that is already needed for the GUI ) because the schema needs it, but then we won't know which file has the typo, so we would have to do some post-process to locate it

:thumbsup: 

>However, this seems like a lot of work, and I don't have time, and I think yours is also limited. What do you think?

All good by me. Once I have worked on the other items at the top of the backlog I will try to tackle this one.

>Edit: Could you have multiple schemas for a single file? To check different parts of it ( per example, developers side and AS variables )

Hmmm, in theory yes, but you would have to validate it twice, running once with a schema, and again with the other schema. But if you want to check different parts (i.e. different keys and objects in YAML) you can try to model that into the Schema Salad configuration. I haven't used Schema Salad in a while, so I will confirm this when I start working on this issue again. Yet another good question! :grimacing: 

Thanks

**Comment by kinow** [comment]
In GitLab by @mcastril on May 31, 2023, 19:08

mentioned in issue auto-ecearth3#1317

**Comment by kinow** [comment]
In GitLab by @eferre1 on Jun 1, 2023, 09:48

Sure, I was thinking also something similar. About the needs, that was one of the complex checks, most of them are more simple (just allowing one value from a list), but then there are others more complex, another example in auto-ecearth would be the processors distribution:

We check that the total amount doesn't exceed the value defined in the jobs (simple check in python, add the values and compare), but then we also check for some specific configurations (depending on some parameters, as in the example), that the value of each component has to be an specific value. Then depending on another parameter, we raise a warning or an error if this specific configuration wasn't met.

Maybe all this checks can be done in some other way here, or in auto-ecearth4 we don't need some of them, but I'm giving the info I have from what we already have implemented :sweat_smile:

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Jul 10, 2023, 10:32

Hey, @kinow! 

This is a follow up on #1066 in which @dbeltrankyl was wondering for a verification on the script. In the final merge request !348, for Autosubmit 3.15b, I have included a very simple verification that just checks if the appended script is the same as the script in which is being added to. It does this by infering the extended script's type by cheking the executable path on the hash bang (**if there no hash bang line present, it does nothing**) with the type inside of Job object (Type.BASH, Type.PYTHON, Type.R, etc).

I feel like this problem is not really related to here, though. In this issue we are working at workflow level valid configurations, whilst the script type is much more of an static one.

**Comment by kinow** [comment]
Hey @manuel-g-castro ,

I thought your merge request was going to add some new configuration keys (extended_tailer, and extended_header?)

If these new configuration keys are added, then wouldn't it make sense to fail the configuration parsing if

```yaml
# File: some_file.yml
EXP:
  EXTENDED_TAILER: example_path_1
  EXTENDED_HEADER: example_path_2
JOBS:
  JOB_A:
    FILE: example.py
    TYPE: python
```

? I think this configuration can be considered invalid without having to inspect/run the code. WDYT?

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Jul 10, 2023, 10:49

What is the Workflow community aiming for? Slimmer or more verbose configuration files. If it is the first, definitely I wouldn't add anything in this checker tool. If the latter, sure! And we save the whole inference of the added script.

**Comment by kinow** [comment]
mentioned in issue startR#177

**Comment by kinow** [comment]
Something that would be nice to be caught by the validator: https://earth.bsc.es/gitlab/digital-twins/de_340/workflow/-/issues/54#note_227522

>BTW, I just found out I wrote THREAD instead of THREADS in the tco1279l137.yml. Could this have caused the wrong spit between nodes?

**Comment by kinow** [comment]
An example of validation in the workflow configuration performed by user provided logic would be:

https://earth.bsc.es/gitlab/es/auto-ecearth3/-/blob/trunk/plugins/config_validator/validator.py

**Comment by kinow** [comment]
Last week we had another issue due to a mistake in the configuration (simple typo). That issue could be easily detected by a validator.

So I spent some time last week and over this weekend thinking about this issue, and searching for what others were doing to validate YAML. I renamed this issue so that it's not about schema salad, as I think I found a few more interesting alternatives.

I believe this issue is doable, and also very interesting for internships and graduate projects. Here's a description with requirements.

## Validation levels

I thought about four levels of validation.

1. YAML syntax
2. Autosubmit Configuration syntax
3. User Configuration syntax
4. Autosubmit Configuration semantics
5. User Configuration semantics

## YAML syntax

This level is validated by `ruamel.yaml` parsers. This is a valid syntax:

```yaml
JOBS:
PLATFORMS:
```

While this is not valid:

```yaml
JOBS:
PLATFORMS
```

## Autosubmit Configuration syntax

At this level, the validator must validate that the syntax of the Autosubmit configuration.

For example, this is valid:

```yaml
JOBS:
  LOCAL_SETUP:
    FILE: LOCAL_SETUP.sh
    PLATFORM: mn4
PLATFORMS:
  MN4:
```

But these are not:

```yaml
JOBS:
  LOCAL_SETUP:
    FILE: 123 # FILE expects a string
    PLATFORM: mn4
PLATFORMS:
  MN4:
```

```yaml
JOBS: True # JOBS expects an object
PLATFORMS:
  MN4:
```

The errors caught here are mainly around the type of field used. For example, the Job provided needs to be an object, but if you provide a boolean, even though that's valid YAML syntax, it's not valid Autosubmit syntax. (Although the concept of syntax and semantics might be a bit busy, it's easier to go with these definitions as there are libraries that use it too.)

## User Configuration syntax

Autosubmit users are able to extend the configuration. Autosubmit has some values that it manages, like `JOBS`, `EXPERIMENT`, and `PLATFORMS`. Users are not allowed to change the syntax of these objects/keys (i.e. users cannot change how a Platform is created).

But other values are allowed, and managed by users. It is possible that the user configuration might be more extensible than the Autosubmit configuration, and not having a way to validate their own configuration means a validator would be only half-useful.

So given some configuration like,

```yaml
JOBS:
PLATFORMS:
RUN:
  SIMULATION:
    MODEL: abc
    ENSEMBLE_ENABLED: True
    # etc.
```

Users must be able to extend the Autosubmit configuration too, giving rules/specification to the Autosubmit validator, used to validate in case a user accidentally types `MODL` instead of `MODEL`, for example.

It's a lot better to catch errors like this when running a command like `autosubmit create` or `autosubmit validate`, instead of waiting until some workflow task/job fails.

## Autosubmit Configuration semantics

At this level, the validator is expected to catch errors like this one:

```yaml
JOBS:
  LOCAL_SETUP:
    FILE: LOCAL_SETUP.sh
    PLATFORM: mn  # invalid platform!
PLATFORMS:
  MN4:
```

The example above is valid YAML. It is also valid Autosubmit syntax, as `JOBS` has a `Job` object, and the `Job` has a `PLATFORM` that is a `String`. But semantically, the selected platform must exist in the `PLATFORMS` object.

## User Configuration semantics

At this level, the Autosubmit validator is expected to allow users to provide extra rules to validate their configuration semantics. For example, if they decide to have a configuration:

```yaml
JOBS:
PLATFORMS:
  MN4:
  LUMI:
RUN:
  SIMULATION:
    VALIDATED_PLATFORMS: ["LUMI", "MN4", "LEONARDO"]
```

Ideally, users must be able to define that `VALIDATOR_PLATFORMS` is not simply a list of strings, but a list of identifiers of Autosubmit platforms. And that that list is expected to contain only valid platforms available in Autosubmit.

In this example, "LEONARDO" is missing, so the validator must raise an error (assuming the user wants it to raise in that case).

**Comment by kinow** [comment]
This is a list of libraries I found that could be used for building the validator as described above.

## Schema Salad

- 

It is Python based, and used by CWL (cwltool and other runners). Schema Salad schemas can be used to validate the configuration at levels 2 and 3, the Autosubmit and User configurations syntax.

Schemas are written in YAML. It supports inheritance, and users could provide extra schemas. There is a certain learning curve to grok how to write and debug schemas.

A bonus is that Schema Salad supports generating HTML/Markdown documentation based on the schema, and also generate code generators (to parse and produce the schema).

Example from their docs,

Data

```yaml
basket:
  - product: bananas
    price: 0.39
    per: pound
    weight: 1
  - product: cucumbers
    price: 0.79
    per: item
    count: 3
```

Schema

```yaml
- name: Product
  doc: |
    The base type for a product.  This is an abstract type, so it
    can't be used directly, but can be used to define other types.
  type: record
  abstract: true
  fields:
    product: string
    price: float

- name: ByWeight
  doc: |
    A product, sold by weight.  Products may be sold by pound or by
    kilogram.  Weights may be fractional.
  type: record
  extends: Product
  fields:
    per:
      type:
        type: enum
        symbols:
          - pound
          - kilogram
      jsonldPredicate: '#per'
    weight: float

- name: ByCount
  doc: |
    A product, sold by count.  The count must be a integer value.
  type: record
  extends: Product
  fields:
    per:
      type:
        type: enum
        symbols:
          - item
      jsonldPredicate: '#per'
    count: int

- name: Basket
  doc: |
    A basket of products.  The 'documentRoot' field indicates it is a
    valid starting point for a document.  The 'basket' field will
    validate subtypes of 'Product' (ByWeight and ByCount).
  type: record
  documentRoot: true
  fields:
    basket:
      type:
        type: array
        items: Product
```

## JSON Schema

- https://json-schema-everywhere.github.io/yaml
- https://github.com/python-jsonschema/jsonschema

JSON Schema is similar to Schema Salad as it also validates at levels 2 and 3. The main differences from Schema Salad are that JSON Schema is an open standard with a community (or at least they are trying to submit it to IETF). There are implementations in different languages, including Python.

I've also seen it mentioned in CWL issues, but I am not sure where/how that's used there. (CWL supports JavaScript expressions, maybe that's where it's used?)

I never used JSON Schema, but reading the docs it looks like it supports vocabularies and schemas, and there is also inheritance between vocabularies. The schemas are defined as JSON, so users would be able to define their own schemas (validating 3).

Example from their docs,

Data

```json
{
  "productId": 1,
  "productName": "A green door",
  "price": 12.50,
  "tags": [ "home", "green" ]
}
```

Schema

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/product.schema.json",
  "title": "Product",
  "description": "A product in the catalog",
  "type": "object"
}
```

## Yamale

Yamale is a library by 23andMe (that DNA service). It is written in Python, and supports reading user provided schemas to parse and validate YAML. Different from Schema Salad and JSON Schema, Yamale does not try to provide a specification to be implemented. It has its own syntax for a schema.

It can be used similarly to Schema Salad or JSON Schema, for levels 2 and 3.

Example from their docs,

Data

```yaml
name: Bill
age: 26
height: 6.2
awesome: True
```

Schema

```yaml
name: str()
age: int(max=200)
height: num()
awesome: bool()
```

It doesn't appear to support inheritance (it supports including schemas, like others), but its syntax is probably the simpler (although probably less flexible than others?).

## iac-validate

`iac-validate` is a client tool, written in Python, but that apparently can also be used programmatically.

>Syntactic validation is done by basic YAML syntax validation (e.g., indentation) and by providing a Yamale schema and validating all YAML files against that schema. Semantic validation is done by providing a set of rules (implemented in Python) which are then validated against the YAML data. Every rule is implemented as a Python class and should be placed in a .py file located in the --rules path.

`iac-validate` is interesting as it delegates the syntactic validation to another library Yamale (could be any of the others though), and then implements a very simple rule system for the semantic validation.

It loads rules from a directory, but probably can be modified to load from Python modules (like plug-ins) as well.

This is the only solution that implements all the levels, 1, 2, 3, and 4.

The rule engine is rather simple, and we could also implement it ourselves.

## Other

For levels 2 and 3, it is probably easier to use an existing library. Especially since if we use one used by Cylc, CWL, ecFlow, etc, then we would be able to improve collaboration and reuse with other workflow managers.

For levels 4 and 5, on the other hand, what we need is a simple rule engine. Here's the signature of the validation function in `iac-validate`:

```py
class Rule:
    id = "101"
    description = "Verify child naming restrictions"
    severity = "HIGH"

    @classmethod
    def match(cls, data):
        results = []
        try:
            for child in data["root"]["children"]:
                if child["name"] == "FORBIDDEN":
                    results.append("root.children.name" + " - " + str(child["name"]))
        except KeyError:
            pass
        return results
```

The rule engine basically loads a YAML file (that has been validated at levels 1, 2, 3?, now at 4 and maybe 5), and calls the `match` function given the YAML data.

Then the `match` function in each rule must be able to validate the semantic/logic. For example, we could have one such as (example code, not tested):

```py
class Rule:
    id = "101"
    description = "Jobs must use existing platforms"
    severity = "HIGH"

    @classmethod
    def match(cls, data):
        results = []
        for job in data["JOBS"]:
            # N.B. no platform equals experiment default platform
            if "PLATFORM" in job and job["PLATFORM"].strip() != "":
                if job["PLATFORM"].strip() not in data["PLATFORMS"]:
                    results.append(f"[ERROR 600x] JOBS.{job['NAME']}.PLATFORM - Platform [{job['PLATFORM'].strip()}] not found!"))
        return results
```

**Comment by kinow** [comment]
I think possible next steps would be:

1. Someone to try to implement a simple validator for level 2, using these three libraries, and provide an assessment (this could be a task for internship or students, cc @mcastril)
2. Verify if this can be extended to support user provided syntax rules/schema (this could be part of the same internship/project, or a separate one)
3. Verify how hard it would be to implement something akin to 23andme's `iac-validator` in pure Python in Autosubmit (less dependencies, better, and it's rather simple), if doable, then proceed to 3 (this is much simpler, than the two previous tasks)
4. Implement an approach using/similar to `iac-validator`, where 2 and 3 are delegated to the syntax validator (levels 2 and 3), and then implement a simple rule engine or use `iac-validator`'s and validate levels 4 and 5

Finally, update documentation, and write tests over the course of the work performed. And that's that :thumbsup: :muscle: 

If we have haste in having validation, it's probably better either @dbeltrankyl or I to work on this. Otherwise, a great project for students or interns. :wave:

**Comment by kinow** [comment]
Another point for JSON Schema is that editors seem to support auto-complete and syntax highlighting based on the JSON Schema configuration.

This website, https://www.schemastore.org/json/, contains over 700 example schemas.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/50174c270cfe318e6137f9225d75fc55/image.png)

This is a very attractive feature, as that could mean that we would be able to validate 2 3, maybe combine with rules or `iac-validate` for 4 5, and also as a bonus be able to auto-complete the autosubmit configuration in VS Code or PyCharm.

**Comment by kinow** [comment]
Bonus bonus point: I read this post today from the Jupyter forum hhttps://discourse.jupyter.org/t/react-json-schema-form-for-jupyterhub-spawner-form/20981

There is a mention there of json schema being used to validate forms in web with react.

> We are exploring the way to use react-jsonschema-form (rjsf) 1 to generate spawner form for our JupyterHub deployment. rjsf prevents the default action on form submit to perform validation checks and we would like to keep those validation checks. 

So the json schemas for jobs, workflows, platforms, etc. could be used in the Autosubmit Web GUI as well.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 29, 2023, 18:39

> If we integrate it after all files are unified, we must store the unify.yml somewhere ( that is already needed for the GUI ) because the schema needs it, but then we won't know which file has the typo, so we would have to do some post-process to locate it

Good point, but I think that if we validate the unified.yml that's already a great step forward, for different reasons:

* With the new (complex) AS4 configuration we give users the opportunity to modify their workflows by only changing one or two files in their local `conf/`. That narrows the search space.

* We don't expect most of the workflows to rely a lot on overloading 

* If we improve the logging of the configuration loading users will be able to track where a parameter is overloaded

**Comment by kinow** [comment]
Related articles:

- https://leebriggs.co.uk/blog/2019/02/07/why-are-we-templating-yaml
- https://cuelang.org/

From CUE's website:

> Just add validation to existing data (CUE, YAML, JSON), reduce boilerplate in large-scale configurations, or both.

**Comment by kinow** [comment]
Note: Maestro uses Yamale: https://maestro-core.readthedocs.io/en/latest/stories.html#user-metadata

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/a7c92d96703c7113b411d775c5d6761d/image.png)

**Comment by kinow** [comment]
mentioned in issue digital-twins/de_340/workflow#370

**Comment by kinow** [comment]
The RO-Crate group is implementing a validator for RO-Crate files. They seem to be experimenting LinkML, which I hadn't heard before, but works with YAML and provides validation too

- https://linkml.io/
- https://linkml.io/linkml/data/validating-data.html
- https://github.com/crs4/rocrate-validator/issues/2

**Comment by kinow** [comment]
In GitLab by @mcastril on Sep 13, 2024, 14:54

mentioned in issue digital-twins/de_340-2/project_management#107

**Comment by dbeltrankyl** [comment]
This would be very good to have at some point

**Comment by ntorqulu** [comment]
Summary of the issue:

The objective is finding a schema validation library for AS unified config file. Is better to validate it all into a single file than a multiple per-file validation as they still contain placeholders, for loops etc... We'll lose provenance, "which file set this" traceability. But the implementation would be less complex and more optimal IMO.

---

Validation layered approach:

### **Level 1**: YAML syntax -> already implemented with `ruamel.yaml`. 
Verifies that the input is a valid yaml and can be parsed.

### **Level 2**: AS config syntax -> need to assess best library
Ensures that the config conforms to the AS configuration schema. E.g. required sections exist, required fields exist, field types, allowed values...

### **Level 3**: User config syntax -> need to assess best library
Validates configuration sections introduced by users. It would need the config files + a corresponding schema describing the restrictions for  the parameters.

### **Level 4**: AS semantics  -> need to assess best library
Validate logical consistency after being parsed. E.g. all the referenced platforms exist, the job dependencies are valid...

### **Level 5**: user-defined rules -> out of scope
Allow users or projects to define its own semantic validation rules.
E.g. maybe the schemas related to destine project should enforce something in specific related to their workflows.
I  would leave that out for now, as it will increase complexity a lot. I write it down it for a future issue.

---

Proposed libraries:

### 1. JSON Schema
I will start with this one. I like that it's an open standard, which gives us interoperability with other tools.
There's a VSCode YAML extension (YAML Language Support by Red Hat) which provides autocomplete, validation while typing, hover documentation, detection of missing required fields, type checking... For inexperienced AS users this would be a nice feature.
The schemas would be YAML (or JSON) files. It addresses levels 2 and 3 but cannot express semantics (level 4, checking that the referenced jobs exist). And there's no framework for user-defined semantic rules (level 5).

For the user schemas, they can define them in a dedicated schema folder in the experiment configuration files:
```
a000
└── schema/
    └── schema.json
```

The schemas could be implemented as pydantic models or json files, I'll see which is better for AS codebase.

**Comment by kinow** [comment]
@ntorqulu 

> I would leave that out for now, as it will increase complexity a lot. I write it down it for a future issue

Agreed. Good call. IMO we just need to make sure whatever direction we go we keep this requirement in mind, so we do not choose a library where this would be difficult or not possible.

About JsonSchema, if we do adopt it, there is also a central repository of schemas. So, if we choose it, we can upload our there, and that would be avail to all consumers of that repo (e.g., vs code, pycharm, etc.).

**Comment by ntorqulu** [comment]
- List of all accepted schemas: https://www.schemastore.org/
- Repo to submit AS schemas: https://github.com/SchemaStore/schemastore

**Comment by kinow** [comment]
Ah, you found it (my brain refused to remember even part of the name... store! 😆 ). That's that one! 

I learned about it from an Australian dev (alex or alexey, I think) who I think was the one that added CWL there.

<img width="651" height="277" alt="Image" src="https://github.com/user-attachments/assets/4e43ce4b-ff90-435b-ab7a-39a1dfb5a6be" />

Pretty useful as I can write CWL with auto-complete in my IDE. Would be nice someday to have autocomplete in Autosubmit too. I believe our users would throw a party for us if that worked well.

**Comment by pablogoitia** [comment]
Hi @kinow. I've just assigned this issue myself. I will turn it into an internship project as soon as I have a free slot 😄

CC @ntorqulu
